### PR TITLE
Reorganize repo to support developer documentation

### PR DESCRIPTION
**Pull request by WardBrian**
This starts the process of using [odoc](https://github.com/ocaml/odoc) to document our internals, similar to what we have for stan-math.

I've uploaded the current version of what gets built here:
https://brianward.dev/stanc3-doc/stanc/index.html

Draft for a few reasons:

- [x] This works kind of like doxygen, so we will want a similar build process to what stan-dev/math uses for their developer documentation. I'll need @serban-nicusor-toptal's help for this. Here's what the odoc repo itself uses: [publish.sh](https://github.com/ocaml/odoc/blob/master/doc/publish.sh)
     - [x] It would be great if we also set up Jenkins to ignore changes outside of test/ and src/ 
- [x] We should clean up a lot of our doc comments to actually have the documentation contain more than just the signatures
- [ ] We may want to move or duplicate parts of the notes folder and github wiki into these docs.

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [x] OR, no user-facing changes were made

## Release notes

Create internal developer documentation

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -567,7 +567,45 @@ pipeline {
...
+                deleteDir()
+                retry(3) { checkout scm }
+                withCredentials([usernamePassword(credentialsId: 'a630aebc-6861-4e69-b497-fd7f496ec46b',
+                                                  usernameVariable: 'GIT_USERNAME', passwordVariable: 'GIT_PASSWORD')]) {
+                    sh """#!/bin/bash
+
+                        set -e
+
+                        dune build @doc
+                        dune build @docgen
```
Comment (WardBrian) [inline Jenkinsfile:583]: We don't want to run `@docgen` for our uses (it is a custom alias for them)

File: Jenkinsfile [hunk 2]
```diff
@@ -567,7 +567,45 @@ pipeline {
...
+
+                        cd ../_build/default/doc
+                        ocaml-mdx test driver.md
+                        cd ../../..
```
Comment (WardBrian) [inline Jenkinsfile:587]: These three are also not useful for us

File: Jenkinsfile [hunk 3]
```diff
@@ -567,7 +567,45 @@ pipeline {
...
+
+                        mkdir -p doc
+                        rsync -av _build/default/doc/html/odoc/ doc/
```
Comment (WardBrian) [inline Jenkinsfile:590]: for us it will be `_build/default/_doc/_html/`

**Comment by WardBrian** [comment]
@serban-nicusor-toptal Something is definitely going wrong during the verify-changes step - my last commit edited many source files but didn't trigger a rebuild

**Comment by serban-nicusor-toptal** [comment]
Checking, it might have something to do with the PR coming from an outside repo.

**Comment by serban-nicusor-toptal** [comment]
Yes, it's using `git diff` see [function](https://github.com/stan-dev/jenkins-shared-libraries/blob/master/src/org/stan/Utils.groovy#L56)

```
+ git checkout master
Already on 'master'
Your branch is up-to-date with 'origin/master'.

Comparing differences between current 3618c13847a91f45c442225d8f2c1a7a315d88d2  and target master

+ git diff 3618c13847a91f45c442225d8f2c1a7a315d88d2 master -- src
+ git diff 3618c13847a91f45c442225d8f2c1a7a315d88d2 master -- test
There aren't any differences in the source code, CI/CD will not run.
```

If `git diff` returns empty when there are no diffs.
I'm debugging but more than sure we need an `if` and to checkout differently for outside PRs

**Comment by serban-nicusor-toptal** [comment]
Yes that's correct

**Comment by WardBrian** [comment]
Good to know. Do you think you've resolved the Jenkins issue? If so this is ready for review

**Pull request merged**